feat: implementation of a basic photo manager

swastikCommits opened the pull request:
<!--
Thank you for contributing to Open Food Facts Explorer!
Please provide a description of your changes below.
-->

## Description

- Added a list of all photos available.
- Added a list of all selected photos (front, ingredient, nutrition, packaging) with a tab system by language.

![image](https://github.com/user-attachments/assets/a83e7a87-80ee-46a6-aafb-67e0cb94e129)

**Currently, I'm waiting for #483 to get merged as it contains the enhanced ImageModal, which will be further used here.**

_After the above PR gets merged, in my upcoming PR, I have made changes locally that will have the ability to Select, Unselect, Rotate and Edit selected photos, which will simultaneously bring the changes to the products/[barcode] page's UI upon editing and saving it in the Photo Manager (**WIP**)_

Fixes #172 

---

## Checklist:

**Author Self-Review:**

- [x] I have performed a self-review of my own code.
- [x] I understand the changes I'm proposing and why they are needed.
- [x] My changes generate no new warnings or errors (linting, console).
- [x] I have made corresponding changes to the documentation (if applicable).

**LLM Usage Disclosure:**
_Please be transparent about the use of AI assistance._

- [ ] If I **did** use an AI Large Language Model, I have reviewed the generated code/text to ensure its accuracy, security, and relevance to the project's context and licensing.

**Triggering Code Review:**

- You can request an AI-powered code review by commenting `/gemini review` on this PR after it's been created.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -0,0 +1,105 @@
+	import type { SelectedImage, Product } from '$lib/api';
+
+	interface Props {
+		product: ReturnType<typeof writable<Product>>;
```
Can't we just say `product: Product`?

swastikCommits replied:
As I'm planning to add image editing functionality in the future, keeping the writable store approach would better in this case

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -0,0 +1,105 @@
+		return ISO6391.getName(code);
+	}
+
+	function getImageUrl(barcode: string, imageName: string): string {
```
Can we move this function to the `lib/...` path, as it is not specific of this component?

swastikCommits replied:
Yeah, I'm moving this function to a relevant path like `src/lib/api/product.ts`

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -0,0 +1,105 @@
+	function getImageUrl(barcode: string, imageName: string): string {
+		const paddedBarcode = barcode.toString().padStart(13, '0');
+		const match = paddedBarcode.match(/^(.{3})(.{3})(.{3})(.*)$/);
+		if (!match) {
+			throw new Error('Invalid barcode format');
+		}
```
I agree with Mr. Gemini here

swastikCommits replied:
Done 👍

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -0,0 +1,105 @@
+		}
+		const path = `${match[1]}/${match[2]}/${match[3]}/${match[4]}`;
+		const image = get(product).images[imageName];
+		if (!image) {
+			return '';
+		}
```
I agree here too

swastikCommits replied:
Done 👍

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -0,0 +1,105 @@
+	];
+
+	function getAllImages(code: string) {
+		const images = [];
+		for (const photoType of photoTypes) {
+			const imageName = `${photoType.id}_${code}`;
+			const imageUrl = getImageUrl($product.code, imageName);
+			if (imageUrl) {
+				images.push({
+					url: imageUrl,
+					alt: `${photoType.label} image for ${getLanguage(code)}`,
+					type: photoType.label
+				});
+			}
+		}
```
Use a functional approach.

```js
photoTypes.map(type => ...).filter(it => it != null).map(...)
```

swastikCommits replied:
Yeah, sure!

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -0,0 +1,105 @@
+			}
+		}
+
+		for (const [key] of Object.entries(get(product).images)) {
+			if (key.endsWith(`_${code}`) && !photoTypes.some((pt) => key === `${pt.id}_${code}`)) {
+				const imageUrl = getImageUrl($product.code, key);
+				if (imageUrl) {
+					images.push({
+						url: imageUrl,
+						alt: `${key} for ${getLanguage(code)}`,
+						type: key.split('_')[0]
+					});
+				}
+			}
+		}
```
Same here, use a functional approach

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -0,0 +1,105 @@
+
+					{#if getAllImages(code).length === 0}
+						<div class="flex w-full items-center justify-center bg-gray-50 p-4">
+							<p class="text-sm text-gray-500">{$_('product.edit.no_photo_available')}</p>
```
Can we remove / replace the `text-gray-500` with something that changes with the theme?

VaiTon reviewed (changes requested):
Thank you! Just some minor comments

swastikCommits commented:
I have made some minor changes to the UI of tabs as well :)
Now it looks better.

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -0,0 +1,100 @@
+			.map((photoType) => {
+				const imageName = `${photoType.id}_${code}`;
+				const imageUrl = getProductImageUrl($product.code, imageName, get(product));
+				return imageUrl
+					? {
+							url: imageUrl,
+							alt: `${photoType.label} image for ${getLanguage(code)}`,
+							type: photoType.label
+						}
+					: null;
```
1. Do not use truthy values
2. Too complicated, use a if-else statement.
```js
if (imageUrl == null) return null;
...
```

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -0,0 +1,100 @@
+			)
+			.map((key) => {
+				const imageUrl = getProductImageUrl($product.code, key, get(product));
+				return imageUrl
+					? {
+							url: imageUrl,
+							alt: `${key} for ${getLanguage(code)}`,
+							type: key.split('_')[0]
+						}
+					: null;
```
Also, maybe first map with getProductImageUrl, filter by null and then map on the dict

swastikCommits commented:
@VaiTon I have performed the changes and refactored it as requested. Additionally, I have added an onMount hook to handle the languages tab in a better way, so that it renders it on the initial load.

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -28,67 +29,93 @@
+	}
+
+	onMount(() => {
+		activeLanguageCode = activeLanguageCode;
```
Why ?

swastikCommits replied:
As I mentioned, the photos were not rendering on the first initial load after the recent changes that I made. They rendered when I switched the languages tab.
So to tackle this, I used the onMount hook, which will force a reactive update when the component mounts, ensuring that the images are correctly loaded on initial render.

VaiTon replied:
Maybe try to use `$state($derived(...`. This is not a pretty solution, if we can manage to avoid it let's do it.

swastikCommits replied:
Yup made the changes.
```
let activeLanguageCode = $state(product.lang || Object.keys(product.languages_codes)[0]);
let currentImages = $state<ProductImage[]>([]);

	$effect(() => {
		currentImages = getImagesForLanguage(activeLanguageCode);
	});
```

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -28,67 +29,93 @@
+		if (input.checked) {
+			const label = input.getAttribute('aria-label');
+			const code = Object.keys(product.languages_codes).find((code) => getLanguage(code) === label);
+			if (code) {
```
Do not use truthy values

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -1,15 +1,16 @@
 	interface Props {
-		product: ReturnType<typeof writable<Product>>;
+		product: Product;
+		barcode: string;
```
Why `barcode` when you can use `product.code`?

swastikCommits replied:
Yeah, you're right, rn I was passing unnecessarily an extra prop for `barcode`, which could have been directly handled by `product.code`

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -0,0 +1,125 @@
+				class="tab"
+				aria-label={getLanguage(code)}
+				checked={code === activeLanguageCode}
+				onchange={handleLanguageChange}
```
Instead of using a raw function here and then using a lot of logic to extract the aria-label, just create a lambda and pass it:

```js
onchange={() => handleLanguageChange(code)}
```

swastikCommits replied:
Yeah thanks this would be a cleaner approach.

VaiTon commented on src/routes/products/[barcode]/edit/PhotoManager.svelte:
```diff
@@ -0,0 +1,125 @@
+	function getImagesForLanguage(code: string): ProductImage[] {
+		const productImages = product.images;
+
+		const standardImages = photoTypes
+			.map((photoType) => {
+				const imageName = `${photoType.id}_${code}`;
+				const imageUrl = getProductImageUrl(product.code, imageName, productImages);
+
+				if (imageUrl == null) {
+					return null;
+				}
+
+				return {
+					url: imageUrl,
+					alt: `${photoType.label} image for ${getLanguage(code)}`,
+					type: photoType.label
+				};
+			})
+			.filter((img): img is ProductImage => img !== null);
+
+		const additionalImages = Object.keys(productImages)
+			.filter(
+				(key) => key.endsWith(`_${code}`) && !photoTypes.some((pt) => key === `${pt.id}_${code}`)
+			)
+			.map((key) => {
+				const imageUrl = getProductImageUrl(product.code, key, productImages);
+
+				if (imageUrl == null) {
+					return null;
+				}
+
+				return {
+					url: imageUrl,
+					alt: `${key} for ${getLanguage(code)}`,
+					type: key.split('_')[0]
+				};
+			})
+			.filter((img): img is ProductImage => img !== null);
+
+		return [...standardImages, ...additionalImages];
```
Can you add some comments?

swastikCommits replied:
Yup, sure!

VaiTon reviewed (commented):
Almost ready! Great work!
